### PR TITLE
Provide gas stipend manually for send(0).

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/977

### CODE REVIEW

**Comment by chriseth** [comment]
Test failures will be resolved after rebasing on top of the payable PR.

**Comment by pirapira** [comment]
Rebased, let's see the test results.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -568,12 +568,17 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 			break;
 		case Location::Send:
 			_functionCall.expression().accept(*this);
-			m_context << u256(0); // do not send gas (there still is the stipend)
+			// Provide the gas stipend manually at first because we may send zero ether.
+			// Will be zeroed if we send more than zero ether.
+			m_context << u256(eth::GasCosts::callStipend);
 			arguments.front()->accept(*this);
```
Comment (pirapira) [inline libsolidity/codegen/ExpressionCompiler.cpp:574]: Does this line push `value` to the stack?  Then the arithmetics below makes sense.

**Comment by pirapira** [comment]
By the way, isn't it the case, sending `1 wei` has the same problem?

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp:574]: Yes, it pushes the argument of the function call `.send(x)` onto the stack (visiting an expression generates code that evaluates the expression and pushes the value onto the stack).

**Comment by chriseth** [comment]
No, sending `1 wei` will trigger the gas stipend mechanism of the EVM, which results in the `call` opcode being more expensive and the callee to receive the gas stipend.

**Comment by pirapira** [comment]
I didn't know EVM calculates the gas of CALL differently if the sent value is zero or non-zero.

**Pull request merged**